common/cuda_hip/factorization/lu_kernels.cpp

```diff
@@ -113,14 +113,17 @@ __global__ __launch_bounds__(default_block_size) void factorize(
     // for each lower triangular entry: eliminate with corresponding row
     for (auto lower_nz = row_begin; lower_nz < row_diag; lower_nz++) {
         const auto dep = cols[lower_nz];
-        // we can load the value before synchronizing because the following
-        // updates only go past the diagonal of the dependency row, i.e. at
-        // least column dep + 1
-        const auto val = vals[lower_nz];
         const auto diag_idx = diag_idxs[dep];
         const auto dep_end = row_ptrs[dep + 1];
         scheduler.wait(dep);
+        // We need to load vals after synchronize.
+        // the next lower_nz might be modified if the dep row has the same col
+        // as next lower_nz's col.
+        const auto val = vals[lower_nz];
         const auto diag = vals[diag_idx];
+        // we need sync to ensure all threads get the data before assigning to
+        // scale.
+        warp.sync();
```

Comment:
I think this still does some superfluous changes that hide the actual fix: The move around `scheduler.wait(dep)` is unnecessary, the only thing that needs to be added is a `warp.sync()` at the top or bottom of the loop (so one iteration doesn't interleave with another iteration)

Reply:
yes, I have mentioned that, but is it not done implicitly by sceduler.wait(dep) together? 
If we would like to make them more clear. them we need two warp.sync() one before and after get val

Reply:
Most of the time is spent in the scheduler.wait function, the additional warp sync doesn't really matter in terms of performance. I think we should also be more precise in our descriptions here, this is mixing up two different levels of dependencies.
- The warp syncs are necessary because of independent thread scheduling and the fact that wait(dep) uses only a single thread to wait
- The values for the current row will only be modified by threads from the same warp, so there is no potential for data races once we add a sync
- The wait function is necessary to synchronize with other warps and their row's values. The fact that it calls syncwarp is an implementation detail we should not rely on, as it might change in the future.